File: android/src/main/kotlin/live/hms/hmssdk_flutter/methods/HMSCameraControlsAction.kt
package live.hms.hmssdk_flutter.methods

import android.content.Context
import io.flutter.plugin.common.MethodCall
import io.flutter.plugin.common.MethodChannel.Result
import live.hms.hmssdk_flutter.HMSErrorLogger
import live.hms.hmssdk_flutter.HMSExceptionExtension
import live.hms.hmssdk_flutter.HMSResultExtension
import live.hms.video.sdk.HMSSDK
import java.io.File
import java.util.*

class HMSCameraControlsAction {

    companion object {
        fun cameraControlsAction(call: MethodCall, result: Result, hmssdk: HMSSDK, context: Context) {
            when (call.method) {
                "is_tap_to_focus_supported" -> isTapToFocusSupported(result, hmssdk)
                "is_zoom_supported" -> isZoomSupported(result, hmssdk)
                "capture_image_at_max_supported_resolution" -> captureImageAtMaxSupportedResolution(call, result, hmssdk, context)
                "is_flash_supported" -> isFlashSupported(result, hmssdk)
                "toggle_flash" -> toggleFlash(result, hmssdk)
                else -> {
                    result.notImplemented()
                }
            }
        }

        // TODO: Complete the method
        private fun isTapToFocusSupported(result: Result, hmssdk: HMSSDK) {
            hmssdk.getLocalPeer()?.let { localPeer ->
                localPeer.videoTrack?.let { localVideoTrack ->
                    localVideoTrack.getCameraControl()?.let { cameraControl ->
                        HMSResultExtension.toDictionary(true, cameraControl.isTapToFocusSupported())
                        return
                    }
                }
            }
            result.success(false)
        }

        // TODO: Complete the method
        private fun isZoomSupported(result: Result, hmssdk: HMSSDK) {
            hmssdk.getLocalPeer()?.let { localPeer ->
                localPeer.videoTrack?.let { localVideoTrack ->
                    localVideoTrack.getCameraControl()?.let { cameraControl ->
                        HMSResultExtension.toDictionary(true, cameraControl.isZoomSupported())
                        return
                    }
                }
            }
            result.success(false)
        }

        /***
         * This is used to capture image at maximum resolution supported by the camera
         * Here we take the file path from user and save the image at that location
         * If flash is already on while calling this method then we turn OFF the flash
         */
        private fun captureImageAtMaxSupportedResolution(call: MethodCall, result: Result, hmssdk: HMSSDK, context: Context) {
            val withFlash: Boolean = call.argument<Boolean>("with_flash") ?: run {
                HMSErrorLogger.returnArgumentsError("withFlash parameter is null in captureImageAtMaxSupportedResolution method")
                false
            }

            // Creating a file at the application directory
            val dir = context.getExternalFilesDir("images")
            val filePath = "$dir/hms_${Date().time}.jpg"
            val imageFile = File(filePath)
            hmssdk.getLocalPeer()?.let { localPeer ->
                localPeer.videoTrack?.let { localVideoTrack ->
                    localVideoTrack.getCameraControl()?.let { cameraControl ->

                        // If flash is already turned ON we turn it OFF before calling this method
                        if (cameraControl.isFlashEnabled()) {
                            cameraControl.setFlash(false)
                        }
                        if (withFlash) {
                            /***
                             * Here we check whether flash is supported by the currently facing camera device
                             */
                            if (cameraControl.isFlashSupported()) {
                                cameraControl.setFlash(true)
                            } else {
                                HMSErrorLogger.logError("captureImageAtMaxSupportedResolution", "Flash is not supported for current facing camera", "Compatibility error")
                            }
                        }
                        cameraControl.captureImageAtMaxSupportedResolution(imageFile) { isSuccess ->
                            // If the method call is successful we return null and in other case
                            // we return the HMSException map
                            if (isSuccess) {
                                result.success(HMSResultExtension.toDictionary(true, filePath))
                            } else {
                                result.success(HMSResultExtension.toDictionary(false, HMSExceptionExtension.getError("Error in capturing image")))
                            }
                            if (withFlash) {
                                cameraControl.setFlash(false)
                            }
                        }

                        return
                    } ?: run {
                        result.success(HMSResultExtension.toDictionary(false, HMSExceptionExtension.getError("cameraControl is null")))
                        return
                    }
                } ?: run {
                    result.success(HMSResultExtension.toDictionary(false, HMSExceptionExtension.getError("local video track is null")))
                    return
                }
            } ?: run {
                result.success(HMSResultExtension.toDictionary(false, HMSExceptionExtension.getError("local peer is null")))
                return
            }
        }

        /***
         * This is used to check whether flash is supported by the current facing camera
         */
        private fun isFlashSupported(result: Result, hmssdk: HMSSDK) {
            hmssdk.getLocalPeer()?.let { localPeer ->
                localPeer.videoTrack?.let { localVideoTrack ->
                    localVideoTrack.getCameraControl()?.let { cameraControl ->
                        // Here we return whether flash is supported or not
                        result.success(HMSResultExtension.toDictionary(true, cameraControl.isFlashSupported()))
                        return
                    } ?: run {
                        result.success(HMSResultExtension.toDictionary(false, HMSExceptionExtension.getError("cameraControl is null")))
                        return
                    }
                } ?: run {
                    result.success(HMSResultExtension.toDictionary(false, HMSExceptionExtension.getError("local video track is null")))
                    return
                }
            } ?: run {
                result.success(HMSResultExtension.toDictionary(false, HMSExceptionExtension.getError("local peer is null")))
                return
            }
        }

        /***
         * This method is used to toggle the flash
         * Only if the flash is supported by the current facing camera
         * If video is muted then the flash light does not turn ON
         */
        private fun toggleFlash(result: Result, hmssdk: HMSSDK) {
            hmssdk.getLocalPeer()?.let { localPeer ->
                localPeer.videoTrack?.let { localVideoTrack ->
                    localVideoTrack.getCameraControl()?.let { cameraControl ->
                        if (cameraControl.isFlashSupported()) {
                            if (cameraControl.isFlashEnabled()) {
                                cameraControl.setFlash(false)
                            } else {
                                cameraControl.setFlash(true)
                            }
                            result.success(HMSResultExtension.toDictionary(true, null))
                            return
                        } else {
                            result.success(HMSResultExtension.toDictionary(false, HMSExceptionExtension.getError("Flash is not supported for current facing camera, Also please ensure camera is turned ON")))
                            return
                        }
                    } ?: run {
                        result.success(HMSResultExtension.toDictionary(false, HMSExceptionExtension.getError("cameraControl is null")))
                        return
                    }
                } ?: run {
                    result.success(HMSResultExtension.toDictionary(false, HMSExceptionExtension.getError("local video track is null")))
                    return
                }
            } ?: run {
                result.success(HMSResultExtension.toDictionary(false, HMSExceptionExtension.getError("local peer is null")))
                return
            }
        }
    }
}

File: ios/Classes/Actions/HMSCameraControlsAction.swift
//
//  HMSCameraControlsAction.swift
//  hmssdk_flutter
//
//  Created by Yogesh Singh on 17/04/23.
//

import Foundation
import HMSSDK

class HMSCameraControlsAction {

    // MARK: - Available Actions

    /// The `cameraControlsAction` function is an action handler for camera controls that can be invoked.
    ///
    /// The function serves as a central point for handling Camera Control actions, routing them to the appropriate implementation.
    ///
    /// - Parameters:
    ///   - call: the Method call received from Flutter framework
    ///   - result: the Result to be passed to Flutter layer
    ///   - hmsSDK: an instance of current HMSSDK
    static func cameraControlsAction(_ call: FlutterMethodCall, _ result: @escaping FlutterResult, _ hmsSDK: HMSSDK?) {

        // Switch statement to handle different camera control method calls that may be invoked
        switch call.method {
        case "capture_image_at_max_supported_resolution":
            captureImageAtMaxSupportedResolution(call, result, hmsSDK)

        case "is_flash_supported":
            isFlashSupported(result, hmsSDK)

        case "toggle_flash":
            toggleFlash(result, hmsSDK)

        // these method calls return FlutterMethodNotImplemented to indicate that the requested functionality is not yet implemented.
        case "is_tap_to_focus_supported", "is_zoom_supported":
            result(FlutterMethodNotImplemented)

        // If the method call does not match any of the above cases, also return FlutterMethodNotImplemented.
        default:
            result(FlutterMethodNotImplemented)
        }
    }

    // MARK: - Capture At Max Resolution

    /// The function captures an image of the local peer's video track at the maximum supported resolution
    ///
    /// It compresses it to JPEG data, saves it to disk, and returns the path of the saved image file as a success result.
    /// The function includes error handling to handle cases where the image capture, compression, or file write fails.
    /// - Parameters:
    ///   - call: the Method call received from Flutter framework
    ///   - result: the Result to be passed to Flutter layer
    ///   - hmsSDK: an instance of current HMSSDK
    static private func captureImageAtMaxSupportedResolution(_ call: FlutterMethodCall, _ result: @escaping FlutterResult, _ hmsSDK: HMSSDK?) {
        guard let localPeer = hmsSDK?.localPeer else {
            result(HMSResultExtension.toDictionary(false, HMSErrorExtension.getError("\(#function) An instance of Local Peer could not be found. Please check if a Room is joined.")))
            return
        }

        guard let localVideoTrack = localPeer.localVideoTrack()
        else {
            result(HMSResultExtension.toDictionary(false, HMSErrorExtension.getError("\(#function) Video Track of Local Peer could not be found. Please check if the Local Peer has permission to publish video & video is unmuted currently.")))
            return
        }

        // Extract the 'with_flash' argument from the Flutter method call arguments
        let arguments = call.arguments as? [AnyHashable: Any]
        let withFlash = arguments?["with_flash"] as? Bool ?? false

        // Capture the image of the Local Peer's Video Track at maximum supported resolution
        localVideoTrack.captureImageAtMaxSupportedResolution(withFlash: withFlash) { image in

            // If capturing the image failed, return an error result
            guard let capturedImage = image else {
                result(HMSResultExtension.toDictionary(false, HMSErrorExtension.getError("\(#function) Could not capture image of the Local Peer's Video Track.")))
                return
            }

            // Compress the captured image to JPEG data
            guard let data = capturedImage.jpegData(compressionQuality: 1.0) else {
                result(HMSResultExtension.toDictionary(false, HMSErrorExtension.getError("\(#function) Could not compress image of the Local Peer's Video Track to jpeg data.")))
                return
            }

            // Write the JPEG data to disk
            let filePath = getDocumentsDirectory().appendingPathComponent("hms_\(getTimeStamp()).jpg")
            do {
                try data.write(to: filePath)
            } catch let error {
                // If writing to disk failed, return an error result
                result(HMSResultExtension.toDictionary(false, HMSErrorExtension.getError("\(#function) Could not write to disk the image data  of the Local Peer's Video Track. \(error.localizedDescription)")))
                return
            }

            // Return the path of the saved image file as a success result
            result(HMSResultExtension.toDictionary(true, filePath.relativePath))
        }
    }

    static private func isFlashSupported(_ result: @escaping FlutterResult, _ hmsSDK: HMSSDK?) {

        guard let localPeer = hmsSDK?.localPeer else {
            result(HMSResultExtension.toDictionary(false, HMSErrorExtension.getError("\(#function) An instance of Local Peer could not be found. Please check if a Room is joined.")))
            return
        }

        guard let localVideoTrack = localPeer.localVideoTrack()
        else {
            result(HMSResultExtension.toDictionary(false, HMSErrorExtension.getError("\(#function) Video Track of Local Peer could not be found. Please check if the Local Peer has permission to publish video & video is unmuted currently.")))
            return
        }

        localVideoTrack.modifyCaptureDevice { device in
            guard let device = device else {
                result(HMSResultExtension.toDictionary(false, HMSErrorExtension.getError("\(#function) Device could not be found")))
                return
            }
            result(HMSResultExtension.toDictionary(true, device.isTorchModeSupported(.on)))
        }
    }

    static private func toggleFlash(_ result: @escaping FlutterResult, _ hmsSDK: HMSSDK?) {

        guard let localPeer = hmsSDK?.localPeer else {
            result(HMSResultExtension.toDictionary(false, HMSErrorExtension.getError("\(#function) An instance of Local Peer could not be found. Please check if a Room is joined.")))
            return
        }

        guard let localVideoTrack = localPeer.localVideoTrack()
        else {
            result(HMSResultExtension.toDictionary(false, HMSErrorExtension.getError("\(#function) Video Track of Local Peer could not be found. Please check if the Local Peer has permission to publish video & video is unmuted currently.")))
            return
        }

        localVideoTrack.modifyCaptureDevice { device in
            guard let device = device else {
                result(HMSResultExtension.toDictionary(false, HMSErrorExtension.getError("\(#function) Device could not be found")))
                return
            }

            if device.isTorchModeSupported(.on) {
                device.torchMode = device.torchMode == .off ? .on : .off
                result(HMSResultExtension.toDictionary(true, nil))
            } else {
                result(HMSResultExtension.toDictionary(false, HMSErrorExtension.getError("\(#function) Flash is not supported for current facing camera, Also please ensure camera is turned ON")))
            }
        }
    }

    // MARK: - Helper Functions

    /// `getDocumentsDirectory` returns the URL of the location on the device's file system where the app can store files that are generated at runtime, such as user-generated content or application state.
    ///
    /// - Returns: the URL of the application's documents directory
    static private func getDocumentsDirectory() -> URL {
        let paths = FileManager.default.urls(for: .documentDirectory, in: .userDomainMask)
        return paths[0]
    }

    /// This `getTimeStamp` function returns the current time as a string in a specific format using a DateFormatter object.
    ///
    /// The function sets the timeStyle property of the DateFormatter to .medium, which formats the time as "hh:mm:ss a" (e.g. "12:34:56 PM").
    ///
    /// - Returns: the string represents the current time in the time zone of the device running the application.
    static private func getTimeStamp() -> String {
        let formatter = DateFormatter()

        formatter.timeStyle = .medium
        return formatter.string(from: Date())
    }
}
